TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/RequestSenderOrchestrator.java

```diff
@@ -13,228 +14,242 @@
-                    if (!sfp.isSuccess()) {
-                        log.atWarn().setCause(sfp.cause()).setMessage(()->"Scheduled future did not successfully run " +
-                                channelInteraction).log();
-                    }
-                });
-            });
+            Optional.ofNullable(schedule.peekFirstItem()).ifPresent(kvp ->
+                    bindNettyScheduleToCompletableFuture(eventLoop, kvp.startTime, kvp.scheduleFuture));
         }), ()->"");
+
+        return workFuture;
     }
 
     private Instant now() {
         return Instant.now();
     }
 
-    private long getDelayFromNowMs(Instant to) {
-        return Math.max(0, Duration.between(now(), to).toMillis());
+    private Duration getDelayFromNowMs(Instant to) {
+        return Duration.ofMillis(Math.max(0, Duration.between(now(), to).toMillis()));
     }
 
-    // TODO - rewrite this - the recursion (at least as it is) is terribly confusing
-    private void sendNextPartAndContinue(NettyPacketToHttpConsumer packetReceiver,
-                                         EventLoop eventLoop, Iterator<ByteBuf> iterator,
-                                         Instant start, Duration interval, AtomicInteger counter,
-                                         StringTrackableCompletableFuture<AggregatedRawResponse> responseFuture) {
+    private DiagnosticTrackableCompletableFuture<String,AggregatedRawResponse>
+    sendSendingRestOfPackets(IPacketFinalizingConsumer<AggregatedRawResponse> packetReceiver,
+                             EventLoop eventLoop,
+                             Iterator<ByteBuf> iterator,
+                             Instant startAt,
+                             Duration interval,
+                             AtomicInteger counter) {
         final var oldCounter = counter.getAndIncrement();
         log.atTrace().setMessage(()->"sendNextPartAndContinue: counter=" + oldCounter).log();
         assert iterator.hasNext() : "Should not have called this with no items to send";
 
-        packetReceiver.consumeBytes(iterator.next());
+        var consumeFuture = packetReceiver.consumeBytes(iterator.next());
         if (iterator.hasNext()) {
-            Runnable packetSender = () -> sendNextPartAndContinue(packetReceiver, eventLoop,
-                    iterator, start, interval, counter, responseFuture);
-            var delayMs = Duration.between(now(),
-                    start.plus(interval.multipliedBy(counter.get()))).toMillis();
-            eventLoop.schedule(packetSender, Math.min(0, delayMs), TimeUnit.MILLISECONDS);
+            return consumeFuture.thenCompose(cf ->
+                    NettyToCompletableFutureBinders.bindNettyScheduleToCompletableFuture(eventLoop,
+                                    Duration.between(now(), startAt.plus(interval.multipliedBy(counter.get()))))
+                    .getDeferredFutureThroughHandle((v,t)-> sendSendingRestOfPackets(packetReceiver, eventLoop,
```

Comment:
Maybe (thenCompose)?  I'm trying that.  A netty schedule operation will fail during shutdown & I don't see why I _wouldn't_ want to fail faster in that situation.  The other call is already in a thenCompose, so the fastness of the failure is already pretty quick, but I think that you're right that it could still be quicker.